### PR TITLE
Introduce ThreadPool and use it for Display()

### PR DESCRIPTION
**Pull request by t2b3**
Rendering the image files to bitmaps is a rather heavy task
that consumes a lot of CPU time (relatively).

Instead of rendering the images one by one,
do this in parallel with threads using the available CPU cores.

The threads are managed by ThreadPool that is introduced in this commit
as the base.

### CODE REVIEW

File: src/threadpool.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
+inline void THROW_OR_ABORT(const std::string &s) {
```
Comment (aardappel) [inline src/threadpool.h:1]: remove this function, turn the error below into an assert

File: src/document.h [hunk 2]
```diff
@@ -545,6 +545,16 @@ struct Document {
         dc.SetBackground(wxBrush(wxColor(Background())));
         dc.Clear();
         if (!rootgrid) return;
+        loopv(i, sys->imagelist) sys->imagelist[i]->trefc = 0;
+        rootgrid->ImageRefCount();
```
Comment (aardappel) [inline src/document.h:549]: Why is this happening on every draw? better to do it at the end of loading

Comment (t2b3) [reply src/document.h]: Ah sorry. I changed that.

File: src/threadpool.h (hunk 1 above)
Comment (t2b3) [reply src/threadpool.h]: Ok I did that.

File: src/system.h [hunk 3]
```diff
@@ -406,6 +406,16 @@ struct System {
 
     done:
 
+        loopv(i, sys->imagelist) sys->imagelist[i]->trefc = 0;
+        doc->rootgrid->ImageRefCount();
+
+        ThreadPool pool(std::thread::hardware_concurrency());   
+        loopv(i, sys->imagelist) {
+            pool.enqueue([](Image *img) {
+                if(img->trefc) img->Display();
```
Comment (aardappel) [inline src/system.h:415]: It be good to put a comment inside `Display` that this runs in multiple threads.. i.e. if anyone ever modifies display, they should not be touching any global resources, and we're assuming that the wx functions its calling are thread safe (do we know?)

Comment (t2b3) [reply src/system.h:418]: Ok I added a comment.

File: src/system.h [hunk 4]
```diff
@@ -41,6 +41,9 @@ struct Image {
     }
 
     wxBitmap &Display() {
+        // This might run in multiple threads in parallel
+        // so this function must not touch any global resources
+        // and callees must be thread-safe.
         if (!bm_display.IsOk()) {
             auto mapitem = imagetypes.find(image_type);
             if (mapitem == imagetypes.end()) return wxNullBitmap;
```
Comment (aardappel) [inline src/system.h:49]: below it accesses the global `sys`, though those functions do not appear to depend on `sys`. So maybe safer to make those function global so it is more obvious what they use.

**Comment by aardappel** [comment]
Thanks!

**Pull request merged**